### ISSUE TITLE
vkGetDeviceProcAddr returning non-NULL for disabled extensions

### ISSUE THREAD

**Issue by rg3igalia** [KhronosGroup/Vulkan-ValidationLayers#1252]
When enabling some validation layers, vkGetDeviceProcAddr returns non-NULL pointers for functions that belong to extensions that have not been enabled in the device. According to the vkGetDeviceProcAddr spec, this should not be happening:

https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#vkGetDeviceProcAddr

I've put together a small reproducer:

https://gist.github.com/rg3igalia/0132cf43bf6eb24189b71da02b6e55dd

It can be compiled defining USE_VALIDATION or not and it prints the returned pointer value.

**Comment by mark-lunarg** [comment]
Looking at this now.  And **THANKS** for the awesome repro case. 🥇

**Comment by mark-lunarg** [comment]
@rg3igalia, this API, `VkTrimCommandPoolKHR `was promoted to core in Vulkan 1.1.  So in 1.1 it is no longer an extension and does not need to be enabled.  If, in your test case, you change `appInfo.apiVersion` from `VK_API_VERSION_1_1 `to `VK_API_VERSION_1_0`, you _will_ get **NULL** since in 1.0 that API is part of the Maintenance1 extension.  Note that the names are aliased in the specification, and in 1.1 the 'KHR' suffix is effectively transparent.  Was this also happening for non-promoted extensions?

**Comment by rg3igalia** [comment]
Thanks for the quick reply and mentioning promotions. I've introduced some changes in the reproducer above so it requests function pointers for vkTrimCommandPool, vkTrimCommandPoolKHR and vkResetQueryPoolEXT. Without enabling the layers, I get output that looks like this:

```
vkTrimCommandPool:    0x7ff769fd9dee
vkTrimCommandPoolKHR: 0
vkResetQueryPoolEXT:  0
```

While with the layers I get output like this one:

```
vkTrimCommandPool:    0x7fdd1fdf09b1
vkTrimCommandPoolKHR: 0x7fdd1fdfb912
vkResetQueryPoolEXT:  0
```

I'm not sure if the behavior observed above is correct or not regarding the first two pointers. There's a note at the beginning of section 3.1 in the spec:

https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#initialization-functionpointers

It reads:

> When extensions are promoted or otherwise incorporated into another extension or Vulkan core version, command aliases may be included. Whilst the behavior of each command alias is identical, the behavior of retrieving each alias’s function pointer is not. A function pointer for a given alias can only be retrieved if the extension or version that introduced that alias is supported and enabled, irrespective of whether any other alias is available.

The way I read it, I think vkTrimCommandPool and vkTrimCommandPoolKHR should be aliases for practically the same thing but vkTrimCommandPool should only be non-NULL if using version 1.1, while vkTrimCommandPoolKHR needs the extension enabled with both 1.0 and 1.1 to be non-NULL.

To give you a bit of context, this bug report comes from a proposed patch to be able to enable the layers in a broader subset of CTS tests. One of the tests is dEQP-VK.api.version_check.entry_points:

https://github.com/KhronosGroup/VK-GL-CTS/blob/372a1fbf47d460b75ec9e2e1de114f82bfd0ac95/external/vulkancts/modules/vulkan/api/vktApiVersionCheck.cpp#L154

The tests assumes vkTrimCommandPoolKHR, among other entry points, will not be available even in 1.1 if the corresponding extension is not enabled, but it fails if you enable the layers for the instance and device being used in the test. If the behavior above is correct, the test would need to be changed, so let me know what you think.

**Comment by jzulauf-lunarg** [comment]
@mark-lunarg  -- looks right.  We're going to have to track whether an extension/alias was actually enabled or with the feature was promoted.  

TRIAGE NOTES: Easiest likely to add a duplicate "without promotion" copy of the InstanceExtensions/DeviceExtensions and modify InitFrom*CreateInfo to take a bool, enum, or tag type argument to control the promote logic that runs there.  Then the GIPA/GDPA setup can look at the "without-promotion" version to return correctly.

We should probably review any of the "enabling extension A, enables alias from extension B" cases as well -- which is apart from the promotion logic and make sure we catch this case (hopefully from the xml)

**Comment by Tony-LunarG** [comment]
Addressed in #1402
